### ISSUE TITLE
is MengeWrapper.GetAgentVelocity working?

### ISSUE THREAD

**Issue by JimSEOW** [MengeCrowdSim/MengeCS#2]
I get true from return but the Vector3 Velocity is { 0, 0, 0}  with the MengeCSExe example

**Comment by MengeCrowdSim** [comment]
Returning a zero velocity is not strictly an error. So, I presume you have knowledge that either

1. the agent you queried is an invalid index (and therefore should've returned false, or
2. the index was valid and you know the agent was moving.

To help you out with this, I'll need some more details:

1) I presume you're doing this in VS 2015?
2) Have you made any modifications to any of the projects/files?
3) Do you have the code that's causing you problems in a branch that I could simply check out to see if I can reproduce?

**Comment by JimSEOW** [comment]
I simply use the [VS2015Project ](https://github.com/MengeCrowdSim/Menge/issues/42)provided in this [branch](https://github.com/curds01/Menge/tree/vs2015)

Compile. Use the MengeCore.dll for MengeCSExe project running the 4Square example as provided. 
Using the same compiled Menge.exe the 4Square example shows velocity for each agent.

**Comment by MengeCrowdSim** [comment]
Ok, that's good to know.  The VS 2015 project was a quick 20-minute hack. It's known not to be complete yet. Tonight I'll be able to sneak an hour into playing with it.  Hopefully, at that point, I'll be able to finish integrating it and testing it against the CS bindings. Thanks for your patience.

**Comment by MengeCrowdSim** [comment]
I hadn't seen that one specifically. It's definitely worth looking into.

I'd originally started with a C++/CLI wrapper but found significant issues with Unity integration. I went with the C-style wrapper because it's a general API that can be accessed in any language that can tie into C-bindings. Still, if I can get a functional output with less effort, so much the better.

**Comment by JimSEOW** [comment]
For cppSharp, just provide a list of header files, MengeCore.lib, it will output Csharp bindings. Look at UrhoSharp as an example.  It is not trivial to start. However, those guys at CppSharp are happy to see more people try and they are willing to help.  The bindings will include C# version of the classes and structures described in the header files. I think this will make writing wrapper simpler. The rest, I think is simply exposing the right way (which is not trivial without really knowing how the codes are organized and classes are inherited)

**Comment by MengeCrowdSim** [comment]
The biggest advantage of the current wrapper is that I can 100% control what people have access through the wrapper. That means, I'm only passing out the rope I want to pass out.  :)

**Comment by JimSEOW** [comment]
> Tonight I'll be able to sneak an hour into playing with it. Hopefully, at that point, I'll be able to finish integrating it and testing it against the CS bindings

Please keep me update. Thank you.

**Comment by MengeCrowdSim** [comment]
Last night didn't go as well as I'd hoped.  Instead of a quick hour to finish it off, I spent three hours. And at the end of it, it was more of a one step forward, two steps back affair.

The 64-bit build in vs 2015 now builds and runs without crashing (although you can no longer dump out screen captures automatically).
But the 32-bit build executable crashes immediately, and linux and vs2013 builds don't build at all. So, not a roaring success.  So, that means I didn't even get a chance to look at its interaction with the C# bindings. 

I'll give it another try tonight.

**Comment by JimSEOW** [comment]
thx for the update, I wait

**Comment by MengeCrowdSim** [comment]
For the record, based on the most current version of the vs2015 branch, I took the 32-blt dll from the VS 2015 build, dropped it into MengeCSExe and I'm getting numbers that all look right.  I've append the first few lines. The values are all within the scope I would expect. So, something weird must be up.  Could you try the latest branch (32-bit build) and tell me what happens?

```
============================
-  Scene root: E:\work\projects\menge_fork\examples\core\4square.
!  Expecting a missing float attribute (rotation) on line 43.  Using default value: 0.
!  Expecting a missing float attribute (rotation) on line 54.  Using default value: 0.
!  Expecting a missing float attribute (rotation) on line 65.  Using default value: 0.
!  Expecting a missing float attribute (rotation) on line 76.  Using default value: 0.
============================
-  Loading behavior from xml: E:\work\projects\menge_fork\examples\core\4square\4squareB.xml
-  Behavior root: E:\work\projects\menge_fork\examples\core\4square
!  Expecting a missing boolean attribute (persistent) on line 5.  Using default value: 0.
!  Expecting a missing boolean attribute (persistent) on line 9.  Using default value: 0.
-  There are 0 registered tasks.

!  No elevation implementation specified.  Using "flat" implementation.
-  For logical time step: 0.1 and 0 sub steps, effective time step is: 0.1New simulator created.
        100 agents
Step 1
        Agent 0 at (-7.268661, -7.19991)
        Agent 1 at (-8.541601, -7.281053)
        Agent 2 at (-9.947639, -7.212485)
        Agent 3 at (-11.15998, -7.211086)
        Agent 4 at (-12.58089, -7.048182)
        Agent 5 at (-7.277556, -8.46586)
        Agent 6 at (-8.630241, -8.63005)
        Agent 7 at (-9.739728, -8.534039)
        Agent 8 at (-11.084, -8.494593)
        Agent 9 at (-12.63195, -8.588489)
        Agent 10 at (-7.236109, -10.00049)
        Agent 11 at (-8.605325, -10.08072)
        Agent 12 at (-9.965252, -10.13864)
        Agent 13 at (-11.30207, -9.950668)
        Agent 14 at (-12.56199, -9.984792)
        Agent 15 at (-7.186831, -11.38874)
        Agent 16 at (-8.730247, -11.30668)
        Agent 17 at (-9.84774, -11.43755)
        Agent 18 at (-11.49629, -11.26351)
        Agent 19 at (-12.59292, -11.18817)
        Agent 20 at (-7.342192, -12.75564)
        Agent 21 at (-8.433341, -12.52613)
        ...
```

And so on...

**Comment by MengeCrowdSim** [comment]
And if I reconfigure MengeCSExe to use the x64 platform target, it also works with the 64-bit MengeCore.dll built in VS2015.

**Comment by curds01** [comment]
The vs2015 branch now has fully functional visual studio 2015 project (32- and 64-bit).  Next up is to update the VS 2013 project and the Linux.  Once that is done, it'll get merged into master.

I ran the VS2013 32- and 64-bit MengeCore.dll (from master) and the VS2015 dlls from this against `MengeCSExe` and they all worked. So, give this a try.

**Comment by curds01** [comment]
@JimSEOW Any word on whether it's working for you?  With the new branch I can't reproduce your problem. So, unless I hear from you, I'll close this when I merge the branch.

**Comment by curds01** [comment]
Try pulling master.  I just pushed a fix.  Stupidly enough, if I hadn't distracted myself with the whole VS 2015 thing, it would've been a 10 minute fix a week ago.  Sorry about that.

**Comment by MengeCrowdSim** [comment]
Closed by #3.
